File: Robot/Java/Polygon.java
package Robot.Java;

import java.util.ArrayList;

import Robot.Java.Helper.ObjectConvertible;
import Robot.Java.Helper.Helper;

class Polygon implements DrawableObject {
	ArrayList<Vector> points = new ArrayList<Vector>();
	
	Polygon() {}
	
	void append(Vector point) {
		points.add(point);
	}
	void move(Vector dv) {
		for (Vector point : points) {
			point.addInPlace(dv);
		}
	}
	void moveToPoint(Vector point) {
		if (points.isEmpty()) {
			return;
		}
		this.move(point.subtract(points.get(0)));
	}
	
	Polygon moved(Vector dv) {
		Polygon poly = new Polygon();
		for (Vector point : points) {
			poly.append(point.add(dv));
		}
		return poly;
	}
	
	ArrayList<Line> getLines() {
		ArrayList<Line> lines = new ArrayList<Line>();
		if (points.isEmpty()) {
			return lines;
		}
		if (points.size() == 1) {
			Vector point = points.get(0);
			lines.add(new Line(point.copy(), point.copy()));
			return lines;
		}
		if (points.size() == 2) {
			lines.add(new Line(points.get(0).copy(), points.get(1).copy()));
			return lines;
		}
		
		Vector currentPoint = points.get(0).copy();
		for (int i = 1; i < points.size(); i += 1) {
			Vector newPoint = points.get(i).copy();
			lines.add(new Line(currentPoint, newPoint));
			currentPoint = newPoint.copy();
		}
		lines.add(new Line(points.get(points.size() - 1).copy(), points.get(0).copy()));
		return lines;
	}
	
	// --- ObjectConvertible
	
	public void convertFromObject(Object object) throws IllegalArgumentException {
		ArrayList<Object> list = Helper.castToArrayListOfObjects(object);
		// !!! can throw
		Helper.throwIllegalErgumentExceptionIfNull("list is null", list);
		
		this.points = new ArrayList<Vector>();
		for (Object obj : list) {
			Vector point = new Vector(0,0);
			// !!! can throw
			point.convertFromObject(obj);
			this.points.add(point);
		}
	}
	
	public Object convertToObject() {
		ArrayList<Object> list = new ArrayList<>();
		for (Vector point : points) {
			list.add(point.convertToObject());
		}
		return list;
	}
	
	// - Helper
	
	public static Polygon fromObject(Object obj) throws IllegalArgumentException {
		Polygon poly = new Polygon();
		// !!! can throw
		poly.convertFromObject(obj);
		return poly;
	}

	// --- Drawable
	
	public MyColor getColor() { return null; }
	public Line getLine() { return null; }
	public Rectangle getRectangle() { return null; }
	public Polygon getPolygon() { return this; }
	public MyImage getImage() { return null; }
}

File: Robot/Polygon.swift

struct Polygon: ObjectConvertible {
	var points = [Vector2D]();
	
	init(points: [Vector2D] = []) {
		self.points = points
	}
	
	mutating func append(_ point: Vector2D) {
		points.append(point);
	}
	func moved(_ dv: Vector2D) -> Polygon {
		var poly = self
		poly.move(dv)
		return poly
	}
	mutating func move(_ dv: Vector2D) {
		for i in 0..<points.count {
			points[i].addInPlace(dv);
		}
	}
	mutating func moveToPoint(_ point: Vector2D) {
		if (points.isEmpty) {
			return;
		}
		self.move(point.subtract(points.get(0)));
	}
	
	func getLines() -> [Line] {
		var lines = Array<Line>();
		if (points.isEmpty) {
			return lines;
		}
		if (points.count == 1) {
			let point = points.get(0);
			lines.add(Line(start: point, end: point));
			return lines;
		}
		if (points.count == 2) {
			lines.add(Line(start: points.get(0), end: points.get(1)));
			return lines;
		}
		
		var currentPoint = points.get(0);
		for i in 1..<points.count {
			let newPoint = points.get(i);
			lines.add(Line(start: currentPoint, end: newPoint));
			currentPoint = newPoint
		}
		lines.add(Line(start: points.get(points.count - 1), end: points.get(0)));
		return lines;
	}
	
	// --- ObjectConvertible
	
	static func convertFromObject(_ object: Object) throws -> Polygon {
		var poly = Polygon()
		try poly.convertFromObject(object)
		return poly
	}
	
	mutating func convertFromObject(_ object: Object) throws {
		guard let list = Helper.castToArrayOfObjects(object) else {
			// !!! can throw
			throw IllegalArgumentException("list is null")
		}
		
		self.points = Array<Vector2D>();
		for obj in list {
			var point = Vector2D(0,0);
			try point.convertFromObject(obj);
			self.points.add(point);
		}
	}
	
	func convertToObject() -> Object {
		return .array(points.map{ $0.convertToObject() })
	}
	
}
